conjure-java-core/src/main/java/com/palantir/conjure/java/util/TypeFunctions.java

```diff
@@ -0,0 +1,208 @@
+import com.palantir.conjure.spec.MapType;
+import com.palantir.conjure.spec.OptionalType;
+import com.palantir.conjure.spec.PrimitiveType;
+import com.palantir.conjure.spec.SetType;
+import com.palantir.conjure.spec.Type;
+import com.palantir.conjure.spec.TypeDefinition;
+import com.palantir.conjure.visitor.TypeDefinitionVisitor;
+import com.palantir.conjure.visitor.TypeVisitor;
+import com.palantir.logsafe.SafeArg;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+public final class TypeFunctions {
+
+    public static boolean isAliasType(Type type) {
+        return type.accept(new IsTypeVisitor() {
+            @Override
+            public Boolean visitReference(com.palantir.conjure.spec.TypeName _value) {
+                return true;
+            }
+        });
+    }
+
+    public static boolean isOptionalBinary(Type type) {
+        return type.accept(new IsTypeVisitor() {
+            @Override
+            public Boolean visitOptional(OptionalType value) {
+                return value.getItemType().accept(TypeVisitor.IS_BINARY);
+            }
+        });
+    }
+
+    public static boolean isBinaryOrOptionalBinary(Type type) {
+        return type.accept(TypeVisitor.IS_BINARY) || isOptionalBinary(type);
+    }
+
+    public static boolean isCollectionType(Type type) {
+        return type.accept(TypeVisitor.IS_LIST) || type.accept(TypeVisitor.IS_SET);
+    }
+
+    // Returns the type that the given alias type refers to. For example, if the input type is defined as
+    // "alias: integer", the returned type will be the type for "integer". The provided type must be an alias
+    // (reference) type.
+    public static Type getAliasedType(Type type, List<TypeDefinition> typeDefinitions) {
+        com.palantir.logsafe.Preconditions.checkArgument(
+                isAliasType(type), "Expected an alias", SafeArg.of("type", type));
+        return getAliasedType(
+                        type.accept(new AbstractTypeVisitor<com.palantir.conjure.spec.TypeName>() {
+                            @Override
+                            public com.palantir.conjure.spec.TypeName visitReference(
+                                    com.palantir.conjure.spec.TypeName value) {
+                                return value;
+                            }
+                        }),
+                        typeDefinitions)
+                .get();
+    }
+
+    public static Optional<Type> getAliasedType(
+            com.palantir.conjure.spec.TypeName typeName, Collection<TypeDefinition> typeDefinitions) {
+        // return type definition for the provided alias type
+        TypeDefinition typeDefinition = typeDefinitions.stream()
+                .filter(typeDef -> {
+                    com.palantir.conjure.spec.TypeName currName = typeDef.accept(TypeDefinitionVisitor.TYPE_NAME);
+                    String currClassName = currName.getPackage() + "." + currName.getName();
```

Comment:
can we compare `TypeName` instead of comparing a string constructed from them?

Reply:
:+1:
I think this has been copied around a lot, that should result in the same behavior without string munging